cpp/include/raft/sparse/neighbors/detail/connect_components.cuh

```diff
@@ -182,46 +194,86 @@ struct LookupColorOp {
-
-  raft::linalg::rowNorm(x_norm.data(), X, n_cols, n_rows, raft::linalg::L2Norm, true, stream);
-
-  raft::distance::fusedL2NN<value_t, raft::KeyValuePair<value_idx, value_t>, value_idx>(
-    kvp,
-    X,
-    X,
-    x_norm.data(),
-    x_norm.data(),
-    n_rows,
-    n_rows,
-    n_cols,
-    workspace.data(),
-    reduction_op,
-    reduction_op,
-    true,
-    true,
-    stream);
+  auto stream = resource::get_cuda_stream(handle);
 
+  value_idx n_components = get_n_components(colors, n_rows, stream);
+
+  // colors_group_idxs is an array containing the *end* indices of each color
+  // component in colors. That is, the value of colors_group_idxs[j] indicates
+  // the start of color j + 1, i.e., it is the inclusive scan of the sizes of
+  // the color components.
+  auto colors_group_idxs = raft::make_device_vector<value_idx, value_idx>(handle, n_components + 1);
+  raft::sparse::convert::sorted_coo_to_csr(
+    colors, n_rows, colors_group_idxs.data_handle(), n_components + 1, stream);
+
+  auto x_norm = raft::make_device_vector<value_t, value_idx>(handle, (value_idx)n_rows);
+  raft::linalg::rowNorm(
+    x_norm.data_handle(), X, n_cols, n_rows, raft::linalg::L2Norm, true, stream);
+  auto kvp_view =
+    raft::make_device_vector_view<raft::KeyValuePair<value_idx, value_t>, value_idx>(kvp, n_rows);
+
+  auto adj     = raft::make_device_matrix<bool, value_idx>(handle, batch_size, n_components);
+  using OutT   = raft::KeyValuePair<value_idx, value_t>;
+  using ParamT = raft::distance::masked_l2_nn_params<red_op, red_op>;
+
+  bool apply_sqrt      = true;
+  bool init_out_buffer = true;
+  ParamT params{reduction_op, reduction_op, apply_sqrt, init_out_buffer};
+
+  size_t n_batches = raft::ceildiv(n_rows, batch_size);
+  for (size_t bid = 0; bid < n_batches; bid++) {
+    size_t batch_offset   = bid * batch_size;
+    size_t rows_per_batch = min(batch_size, n_rows - batch_offset);
+
+    auto X_view = raft::make_device_matrix_view<const value_t, value_idx>(
+      X + batch_offset * n_cols, rows_per_batch, n_cols);
+
+    auto x_norm_view = raft::make_device_vector_view<value_t, value_idx>(
+      x_norm.data_handle() + batch_offset, rows_per_batch);
+    auto mask_op = [colors,
+                    n_components = raft::util::FastIntDiv(n_components),
+                    batch_offset] __device__(value_idx idx) {
+      value_idx row = idx / n_components;
+      value_idx col = idx % n_components;
+      return colors[batch_offset + row] != col;
+    };
+    auto adj_view = raft::make_device_matrix_view<bool, value_idx>(
+      adj.data_handle(), rows_per_batch, n_components);
+    raft::linalg::map_offset(handle, adj_view, mask_op);
+
+    raft::distance::masked_l2_nn<value_t, OutT, value_idx, red_op, red_op>(
+      handle,
+      params,
+      X_view,
+      X_view,
+      x_norm_view,
+      x_norm_view,
+      adj_view,
+      raft::make_device_vector_view(colors_group_idxs.data_handle() + 1, n_components),
+      kvp_view);
+  }
```

Comment:
Please use resource::get_thrust_policy() from  raft/core/resource/thrust_policy.hpp instead of rmm::exec_policy.